controlplane/kubeadm/internal/machinefilters/machine_filters.go

```diff
@@ -197,3 +205,114 @@ func ControlPlaneSelectorForCluster(clusterName string) labels.Selector {
+		if clonedFromName != kcp.Spec.InfrastructureTemplate.Name ||
+			clonedFromGroupKind != kcp.Spec.InfrastructureTemplate.GroupVersionKind().GroupKind().String() {
+			return false
+		}
+		return true
+	}
+}
+
+// MatchesKubernetesVersion returns a filter to find all machines that match a given Kubernetes version.
+func MatchesKubernetesVersion(kubernetesVersion string) Func {
+	return func(machine *clusterv1.Machine) bool {
+		if machine == nil {
+			return false
+		}
+		return *machine.Spec.Version == kubernetesVersion
+	}
+}
+
+// MatchesKubeadmBootstrapConfig checks if machine's KubeadmConfigSpec is equivalent with KCP's KubeadmConfigSpec.
+func MatchesKubeadmBootstrapConfig(ctx context.Context, c client.Client, kcp controlplanev1.KubeadmControlPlane, cluster clusterv1.Cluster) Func {
+	return func(machine *clusterv1.Machine) bool {
+		bootstrapRef := machine.Spec.Bootstrap.ConfigRef
+		if bootstrapRef == nil {
+			// Missing bootstrap reference should not be considered as unmatching.
+			return true
+		}
+
+		bootstrapObj := &bootstrapv1.KubeadmConfig{}
+		if err := c.Get(ctx, client.ObjectKey{Name: bootstrapRef.Name, Namespace: machine.Namespace}, bootstrapObj); err != nil {
+			// Return true here because failing to get KubeadmConfig should not be considered as unmatching.
+			return true
+		}
+
+		kcpConfigSpecLocal := kcp.Spec.KubeadmConfigSpec.DeepCopy()
+		if bootstrapObj.Spec.InitConfiguration == nil {
+			kcpConfigSpecLocal.InitConfiguration = nil
+		}
+		if bootstrapObj.Spec.JoinConfiguration == nil {
+			kcpConfigSpecLocal.JoinConfiguration = nil
+		}
+
+		// If machine's ClusterConfiguration is not nil, then semantic comparison needs to be done before comparing it with KCP
+		// ClusterConfiguration as some fields may be missing such as ControlPlaneEndpoint, Networking.
+		if bootstrapObj.Spec.ClusterConfiguration == nil {
+			// Machines that have ClusterConfigurationAnnotation will have to match with KCP ClusterConfiguration.
+			machineClusterConfigStr, ok := machine.GetAnnotations()[clusterv1.ClusterConfigurationAnnotation]
+			if ok {
+				machineClusterConfig := &kubeadmv1.ClusterConfiguration{}
+				err := json.Unmarshal([]byte(machineClusterConfigStr), machineClusterConfig)
+				// ClusterConfiguration annotation is not correct, only solution is to rollout.
+				if err != nil {
+					return false
+				}
+				if !reflect.DeepEqual(machineClusterConfig, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration) {
+					return false
+				}
+			} else {
+				// ClusterConfigurationAnnotation is missing, that means the machine is either old or adopted. Do not initiate rollout.
+				// This is a known limitation, user needs to trigger rollout using UpgradeAfter or changing some other field in KCP such as Kubernetes version
+				// as older machines (or adopted ones) will not detect with ClusterConfiguration changes in KCP.
+				kcpConfigSpecLocal.ClusterConfiguration = nil
+			}
```

Comment:
Do we need the `else` block or can we reset the `ClusterConfiguration` to nil regardless?

Reply:
I also removed this check `if bootstrapObj.Spec.ClusterConfiguration == nil {}` on L284. Previously, if this is not nil, we were relying on comparison with Machine's ClusterConfiguration.

Now the logic: compare ClusterConfiguration with annotation, if annotation is not set, set KCP ClusterConfiguration to nil to avoid any comparison with Machine's ClusterConfiguration. This way we don't have a second path for the machine initialize the control plane and also adopted machines ClusterConfiguration will not be checked because we say changes there will not trigger rollout.